lib: list fixes and enhancements

garrett opened the pull request:
Addresses:

- #10676 
- #10672
- #10689 
- dark lines in items in lists-in-lists https://github.com/cockpit-project/cockpit/pull/10335#issuecomment-436326752

And a few additional issues I found along the way.

garrett commented:
I squashed everything into separate but related commits. Should I change commit messages? (Using `lib/list`; should it be something else? Should I use other capitalization? Etc.)

I've included "Closed #" where appropriate.

Also, I specifically did _not_ include the overflow/scrolly WIP I have been working on. It needs more work and cross-browser testing and should be a separate PR.

andreasn commented:
I still get the label overlap (#10689) with this applied.
![screenshot from 2018-11-29 12-45-54](https://user-images.githubusercontent.com/209895/49219853-d3649200-f3d4-11e8-9a23-5e2ada801ba3.png)

And the misaligned headers (#10672)
![screenshot from 2018-11-29 12-47-14](https://user-images.githubusercontent.com/209895/49219881-eb3c1600-f3d4-11e8-834c-d5f0a0c8ae2a.png)

Maybe I didn't apply it correctly? I could try nuking the cockpit folder /usr/share

garrett commented:
@andreasn: This PR doesn't attempt to solve the label overlap. To do that properly, we need to use CSS Grid. We could use tables, but that's a mess and also incorrect to use for layout. Those are the only two methods to have dynamic length for text that lines up. I intend on porting it to grid (which should be easy-peasy for this), but I'm tackling a few other things first. :wink:

I did add in a fix for the panel with floats though, so there's padding at the bottom on the overview tab now.

As for the misaligned headers, I rebased this branch and added some fixes since your comment. (Thanks again for mentioning it!) Hopefully it's solved now?

garrett commented:
I'm working on adding CSS-grid-based layout for forms for #10703. The intention is to support modals first. But after looking at your screenshot again, I think we could actually re-use the exact same CSS here too.

andreasn commented:
I get a regression on the containers page with this branch applied:
![screenshot from 2018-11-30 14-13-01](https://user-images.githubusercontent.com/209895/49291293-21979500-f4aa-11e8-8ccd-1fa524b5bd91.png)

The pane doesn't display a double border on master, but it does with this branch.

andreasn commented:
Just a quick note that this PR doesn't fix all the issues mentioned in https://github.com/cockpit-project/cockpit/issues/10672#issuecomment-441618189 and https://github.com/cockpit-project/cockpit/issues/10689, just so that doesn't accidentally sneak into the commit message.
For #10672, it can be split out into a separate issue, so don't worry about fixing it all in one go.

garrett commented on pkg/machines/components/vmDisksTab.jsx:
```diff
@@ -119,7 +119,7 @@ const VmDisksTab = ({ idPrefix, vm, disks, actions, renderCapacity, notification
                             idPrefixRow,
                         });
-                        columns.push(<div>{removeDiskAction}</div>);
+                        columns.push(<React.Fragment>{removeDiskAction}</React.Fragment>);
```
It probably doesn't need to be wrapped. I just changed the `<div>` to a `<React.Fragment>` as it shouldn't be wrapped in the resulting HTML.

What's the best way to do this in React? Is this correct?

```js
columns.push({removeDiskAction});
```

martinpitt replied:
Just `colums.push(removeDiskAction);`

garrett commented:
@andreasn: The Containers page is at fault: It nests `listing-ct-body` inside itself and "bad things happen". I have a fix in a commit that I'm testing right now.

In a few minutes, I'll force push an update to this PR that includes the fix, a rebase, Closes #, and commit message rewording.

garrett commented:
@andreas, @martinpitt: Everything the two of you brought up should be fixed, and the branch has been rebased.. Please let me know if there's anything else needed.

andreasn commented on pkg/lib/listing.less:
```diff
@@ -180,12 +186,26 @@ tbody.open td.listing-ct-toggle i:before {
+        margin-bottom: -7px;
+    }
+    > .btn-control-ct {
+        /* Cockpit's flavor embeggins the button by 8px */
```
What does this code comment mean?
Is _embiggins_ a Quiddish thing? https://wordability.net/tag/embeggin/
Flavor of what?

garrett replied:
Embiggen is an official word in the Merriam Websters dictionary meaning "to make larger".
https://www.merriam-webster.com/dictionary/embiggen

It's originally from The Simpsons, from a 1996 episode which also coined the word "cromulent" (meaning "perfectly acceptable").

Since then, "A noble spirit embiggens the smallest man" has been added to the intro, at the bottom of the Jebediah Springfield statue. (Probably added when they went to HD? I don't know exactly.)

I accidentally swapped the e and i. Oops.

garrett replied:
It should be:

```suggestion
        /* Cockpit's flavor embiggens the button by 8px */
```

It's an easter egg in the source, basically. :wink:

garrett replied:
Changed to:
```css
        /* Adapt for Cockpit's modified button, which is larger by 8px */
```

garrett commented:
Removed a set of curlies (`{}`) and force-pushed with the change. (Almost missed the comment about them not being needed!)
